src/ui/carousel/carousel.ts

```diff
@@ -0,0 +1,506 @@
+  private readonly condition: () => boolean;
+  private readonly container: HTMLElement;
+  private readonly slides: HTMLElement[];
+  private readonly transition: Transition;
+  private readonly allowLooping: boolean;
+  private readonly onTransitionCallbacks: Array<(carousel: Carousel) => void>;
+  private readonly onDisposeCallbacks: Array<(carousel: Carousel) => void>;
+  private readonly slideCssClasses: DefaultMap<HTMLElement, Set<string>>;
+  private readonly raf: Raf;
+  private readonly synchronizer: CarouselSynchronizer;
+  private transitionTarget: TransitionTarget;
+  private lastActiveSlide: HTMLElement;
+
+  /**
+   * @param container Parent element of slides.
+   * Element that acts as the carousel, in which all slides are contained.
+   * Does not need to be the direct parent.
+   *
+   * @param slides HTMLElements containing slide content.
+   *
+   * @param condition Under what conditions the carousel should run
+   * @param onTransitionCallbacks Functions run when the active slide changes.
+   * @param onDisposeCallbacks Functions run when the carousel is disposeed.
+   * @param activeCssClass Class to apply to active slide.
+   * @param beforeCssClass Class to apply to slides before active slide.
+   * @param afterCssClass Class to apply to slides after active slide.
+   * @param allowLooping Whether the carousel should be allowed to loop.
+   * @param transition How the Carousel should transition.
+   * Please see the transitions folder inside the carousel folder for options.
+   */
+  constructor(
+      container: HTMLElement,
+      slides: HTMLElement[],
+      {
+        condition = () => true,
+        onTransitionCallbacks = [],
+        onDisposeCallbacks = [],
+        activeCssClass = DefaultCssClass.ACTIVE_SLIDE,
+        beforeCssClass = DefaultCssClass.BEFORE_SLIDE,
+        afterCssClass = DefaultCssClass.AFTER_SLIDE,
+        allowLooping = true,
+        transition = null
+      }: CarouselOptions = {}
+  ) {
+    if (slides.length < 1) {
+      throw new Error('Cannot start carousel without slides');
+    }
+    this.raf = new Raf(() => this.loop());
+    this.activeCssClass = activeCssClass;
+    this.beforeCssClass = beforeCssClass;
+    this.afterCssClass = afterCssClass;
+    this.allowLooping = allowLooping;
+    this.condition = condition;
+    this.container = container;
+    this.lastActiveSlide = null;
+    this.onTransitionCallbacks = onTransitionCallbacks;
+    this.onDisposeCallbacks = onDisposeCallbacks;
+    this.slides = slides;
+    this.transition = transition !== null ? transition : new CssClassesOnly();
+    this.transitionTarget = null;
+    this.synchronizer = CarouselSynchronizer.getSingleton(this);
+
+    this.slideCssClasses =
+        DefaultMap.usingFunction<HTMLElement, Set<string>>(
+            () => new Set<string>());
+
+    /** Mapped Set caches to avoid allocation churn */
+    this.activeCssClassSet = new Set([activeCssClass]);
+    this.beforeCssClassMap =
```

Comment:
I'm sorry, I don't follow. This caching is designed to set the value once.

These are to prevent allocating memory for the same objects multiple times, so they reduce overall memory.

I need the sets so items can be quickly checked for membership, I could do it with arrays if they are less memory intensive than sets, but we're giving up O(1) access for O(n) access.

What I can do is drop the activeCssClassSet but:
- It would prevent the same method from being used for changing the active slides classes that is used to change the other slides classes due to the different data structure.
- A set would still need to be created for it when setting this.slideCssClasses.

Reply:
I took some time to trace what's going on, and my understanding is that all these sets are being allocated so we know which elements have classes of `active` vs `before before--{index}` and `after after--{index}`.

Sets are extremely memory intensive and I think you can simplify this logic so that it's more readable. You can do something like:

on active element: remove "before" and "after" classes on the element. apply "active"
on inactive elements: remove "active" class. apply either "before" or "after" class.
on all elements: set something like `data-index="{index}"`

This would avoid us having to store and figure out what classes we've already set on the slide.

Reply:
Done. I hadn't considered using an attribute for specifying the index, good call.